test/e2e/autoscaler.go

```diff
@@ -0,0 +1,248 @@
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+// AutoscalerSpecInput is the input for AutoscalerSpec.
+type AutoscalerSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+	ControlPlaneWaiters   clusterctl.ControlPlaneWaiters
+
+	// Flavor, if specified is the template flavor to be used for this test.
+	// Note:
+	//  - the file creating the service account to be used by the autoscaler when connecting to the management cluster
+	//    - must be named "autoscaler-to-workload-management.yaml"
+	//    - must deploy objects in the $CLUSTER_NAMESPACE
+	//    - must create a service account with name "cluster-$CLUSTER_NAME" and the RBAC rules required to work.
+	//    - must create a secret with name "cluster-$CLUSTER_NAME-token" and type "kubernetes.io/service-account-token".
+	//  - the file creating the autoscaler deployment in the workload cluster
+	//    - must be named "autoscaler-to-workload-workload.yaml"
+	//    - must deploy objects in the cluster-autoscaler-system namespace
+	//    - must create a deployment named "cluster-autoscaler"
+	//    - must run the autoscaler with --cloud-provider=clusterapi,
+	//      --node-group-auto-discovery=clusterapi:namespace=${CLUSTER_NAMESPACE},clusterName=${CLUSTER_NAME}
+	//      and --cloud-config pointing to a kubeconfig to connect to the management cluster
+	//      using the token above.
+	//    - could use following vars to build the management cluster kubeconfig:
+	//      $MANAGEMENT_CLUSTER_TOKEN, $MANAGEMENT_CLUSTER_CA, $MANAGEMENT_CLUSTER_ADDRESS
+	Flavor                 *string
+	InfrastructureProvider string
+	AutoscalerVersion      string
+}
+
+// AutoscalerSpec implements a test for the autoscaler, and more specifically for the autoscaler
+// being deployed in the workload cluster.
+func AutoscalerSpec(ctx context.Context, inputGetter func() AutoscalerSpecInput) {
+	var (
+		specName         = "autoscaler"
+		input            AutoscalerSpecInput
+		namespace        *corev1.Namespace
+		cancelWatches    context.CancelFunc
+		clusterResources *clusterctl.ApplyClusterTemplateAndWaitResult
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(input.InfrastructureProvider).ToNot(BeNil(), "Invalid argument. input.InfrastructureProvider can't be empty when calling %s spec", specName)
```

Comment:
I think for 99% of infra providers this field is not required (As far as I can tell only for kubemark)

Reply:
(we would use clusterctl.DefaultInfrastructureProvider if it is empty, compare with quick_start.go)